ssl/quic/quic_impl.c

```diff
@@ -1555,32 +1668,182 @@ int ossl_quic_conn_on_handshake_confirmed(QUIC_CONNECTION *qc)
-            reason != NULL ? reason : "");
-    // TERMINATING - CLOSING
-    // TERMINATING - DRAINING
-    // TERMINATED
-    /* TODO */
+    switch (qc->state) {
+        default:
+        case QUIC_CONN_STATE_IDLE:
+            qc->terminate_cause = *tcause;
+            csm_on_terminating_timeout(qc);
+            break;
+
+        case QUIC_CONN_STATE_ACTIVE:
+            qc->state = tcause->remote ? QUIC_CONN_STATE_TERMINATING_DRAINING
+                                       : QUIC_CONN_STATE_TERMINATING_CLOSING;
+            qc->terminate_cause = *tcause;
+            /* TODO: Calculate deadline properly */
+            qc->terminate_deadline
+                = ossl_time_add(ossl_time_now(), ossl_ms2time(2000));
+
+            if (!tcause->remote) {
+                OSSL_QUIC_FRAME_CONN_CLOSE f = {0};
+
+                /* best effort */
+                f.error_code = qc->terminate_cause.error_code;
+                f.frame_type = qc->terminate_cause.frame_type;
+                f.is_app     = qc->terminate_cause.app;
+                ossl_quic_tx_packetiser_schedule_conn_close(qc->txp, &f);
+                qc->conn_close_queued = 1;
+            }
+            break;
+
+        case QUIC_CONN_STATE_TERMINATING_CLOSING:
+            if (tcause->remote)
+                qc->state = QUIC_CONN_STATE_TERMINATING_DRAINING;
+
+            break;
+
+        case QUIC_CONN_STATE_TERMINATING_DRAINING:
+            /* We remain here until the timout expires. */
+            break;
+
+        case QUIC_CONN_STATE_TERMINATED:
+            /* No-op. */
+            break;
+    }
+}
+
+/*
+ * ossl_quic_conn_on_remote_conn_close {{{2
+ * -----------------------------------
+ */
+void ossl_quic_conn_on_remote_conn_close(QUIC_CONNECTION *qc,
+                                         OSSL_QUIC_FRAME_CONN_CLOSE *f)
+{
+    QUIC_TERMINATE_CAUSE tcause = {0};
+
+    if (!is_active(qc))
+        return;
+
+    tcause.remote     = 1;
+    tcause.app        = f->is_app;
+    tcause.error_code = f->error_code;
+    tcause.frame_type = f->frame_type;
+
+    csm_start_terminating(qc, &tcause);
+}
+
+/*
+ * ossl_quic_conn_raise_protocol_error {{{2
+ * -----------------------------------
+ *
+ * This function is the master function which should be called in the event of a
+ * protocol error detected by us. We specify a QUIC transport-scope error code
+ * and optional frame type which was responsible. The reason string is not
+ * currently handled. If the connection has already terminated due to a previous
+ * protocol error, this is a no-op; first error wins.
+ */
+void ossl_quic_conn_raise_protocol_error(QUIC_CONNECTION *qc,
+                                         uint64_t error_code,
+                                         uint64_t frame_type,
+                                         const char *reason)
+{
+    QUIC_TERMINATE_CAUSE tcause = {0};
+
+    tcause.error_code = error_code;
+    tcause.frame_type = frame_type;
```

Comment:
Shouldn't we call `ERR_raise` somewhere in this function? If so we should consider how we pass through the originating line number data.